### ISSUE TITLE
Having trouble copying shallow parameters to deeper networks

### ISSUE THREAD

**Issue by olivia2222** [CORE-Robotics-Lab/ProLoNets#9]
I ran the code with the command `python gym_runner.py -a prolo -deep -e 1500 -env lunar`
When I debugged the code, I found that the parameters could not be copied successfully. 

https://github.com/CORE-Robotics-Lab/ProLoNets/blob/b1f138ed3520e623a84f2a7d5b1969fa45845700/agents/prolonet_agent.py#L219C1-L237C84

https://github.com/CORE-Robotics-Lab/ProLoNets/blob/b1f138ed3520e623a84f2a7d5b1969fa45845700/opt_helpers/ppo_update.py#L175C17-L193C90

Did I miss any detail? Is this correct? Thanks a lot.

**Comment by andrewsilva9** [comment]
Hi @olivia2222,

I double checked and I'm seeing the same issue. I think this is unfortunately down to PyTorch upgrades and a lack of backwards compatibility with this work (we tested/developed this code a few years ago). With upgrades to PyTorch, the way that `nn.Parameter` objects perform copying seems to have changed, and overall the implementation of ProLoNets is out-of-date. Developing a newer ProLoNet implementation with `nn.Linear` layers _is_ possible and is probably a good way to do it, but unfortunately I do not have the time to do a proper rewrite of this codebase right now.

I've implemented a demonstration of how it might work with this branch: https://github.com/CORE-Robotics-Lab/ProLoNets/tree/deepen_with_linears but the branch does not work properly (for example, gradients aren't flowing to the comparators, and results are looking considerably worse than what I expect). I am not able to do a proper investigation/remediation at this time, but hopefully this is a useful demonstration/starting point for you.

**Comment by mgombolay** [comment]
Thanks, Andrew!On Mar 4, 2024, at 10:04 PM, andrewsilva9 ***@***.***> wrote:﻿
Hi @olivia2222,
I double checked and I'm seeing the same issue. I think this is unfortunately down to PyTorch upgrades and a lack of backwards compatibility with this work (we tested/developed this code a few years ago). With upgrades to PyTorch, the way that nn.Parameter copying seems to have changed, and overall the implementation of ProLoNets is out-of-date. Developing a newer ProLoNet implementation with nn.Linear layers is possible and is probably a good way to do it, but unfortunately I do not have the time to do a proper rewrite of this codebase right now.
I've implemented a demonstration of how it might work with this branch: https://github.com/CORE-Robotics-Lab/ProLoNets/tree/deepen_with_linears but the branch does not work properly (for example, gradients aren't flowing to the comparators, and results are looking considerably worse than what I expect). I am not able to do a proper investigation/remediation at this time, but hopefully this is a useful demonstration/starting point for you.

**Comment by olivia2222** [comment]
Thanks for your reply. I'm sorry for not being clear earlier. The PyTorch version I used for running the code is the same as the one specified in `python38.txt` file  (I also referenced other library versions from `python38.txt`). In the case, could this issue be due to the PyTorch version incompatibility?

I have also made some attempts to solve this issue. And by replacing the code `agent_in/self.deeper_action/value_network.layers/comparators[weight_index].data = new_act/value_weight/comp` with `agent_in/self.deeper_action/value_network.layers/comparators[weight_index].data.copy_(new_act/value_weight/comp)`, the shallow params can be copied to the deeper network correctly, but the result is worse than before the replacement.
![image](https://github.com/CORE-Robotics-Lab/ProLoNets/assets/74092371/79b25589-b0bd-4584-952f-3b058a3594e1)
I ran the code with the command `python gym_runner.py -a prolo -deep -e 1500 -env lunar --reproduce` so the seeds are the same. The orange line represents the results from code`agent_in/self.deeper_action/value_network.layers/comparators[weight_index].data = new_act/value_weight/comp`  , while the blue line represents the results from code `agent_in/self.deeper_action/value_network.layers/comparators[weight_index].data.copy_(new_act/value_weight/comp)`.  The performance is averaged over 5 seeds.

**Comment by andrewsilva9** [comment]
Thanks for more info! 

> In the case, could this issue be due to the PyTorch version incompatibility?

I suspect it is still the source of the issue, as that requirements file was generated some time after we published the work (and it was an attempt to bring some reproducibility to the repo, as we hadn’t included random seeds or exact version matching when we originally published this). The original PyTorch version was something older (remnants of a 1.3.0 version are commented out in the `gym_requirements.txt` file). 

However if your solution is also working, then that seems like a fine option, though again I can’t guarantee it will work exactly as I might expect! The dynamic growth piece _can_ hurt performance in the short term, as you’re adding randomness into the network. But you’re also adding capacity to the network, which _should_ help performance longer-term.

**Comment by olivia2222** [comment]
Thanks for your patience! I tried using PyTorch version 1.3.0 but unfortunately the shallow params still could not be copied to the deeper networks by running `agent_in.deeper_action.layers[weight_index].data = new_act_weight`. I will continue to try to find a solution. 

Besides, I'd like to ask why we need to copy the shallow parameters to the deeper network before updating parameters in `ppo_update.py`? I couldn't find a corresponding explanation in the paper.

**Comment by andrewsilva9** [comment]
Got it. Thanks for following up on the older PyTorch!

I believe the parameter overwrite in the `ppo_update.py` was to try to keep the deeper-layer weights from diverging too far from the original tree. If there were too many `ppo_epoch` steps, then the deeper decision nodes/leaves could drift far from the shallow network during the update. If we then copied those deeper nodes/leaves over, they might have been updated to fit under a tree that drifted too far from the original. However, I don't remember doing extensive testing on this, so it may not be necessary (especially because we stuck with only a few `ppo_epoch` steps).

**Comment by olivia2222** [comment]
OK, I get it. Thank you for your time! I'm gonna close this issue.

**Comment by olivia2222** [comment]
OK, I get it. Thank you for your time! I'm ganna close this issue.

**Comment by olivia2222** [comment]
OK, I get it. Thank you for your time! I'm gonna close this issue.
